plugins/event/sqs/sqs.go

```diff
@@ -0,0 +1,248 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: BUSL-1.1
+
+package sqs
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/aws/aws-sdk-go/service/sqs"
+	"github.com/hashicorp/go-secure-stdlib/awsutil"
+	"github.com/hashicorp/vault/sdk/event"
+	"github.com/hashicorp/vault/sdk/logical"
+	"github.com/hashicorp/vault/version"
+	"github.com/mitchellh/mapstructure"
+)
+
+var _ event.EventSubscriptionPlugin = (*sqsBackend)(nil)
+
+const pluginName = "sqs"
+
+// ErrQueueRequired is returned if the required queue parameters are not present.
+var ErrQueueRequired = errors.New("queue_name or queue_url must be specified")
+
+// New returns a new instance of the SQS plugin backend.
+func New() event.EventSubscriptionPlugin {
+	return &sqsBackend{
+		connections: map[string]*sqsConnection{},
+	}
+}
+
+type sqsBackend struct {
+	connections map[string]*sqsConnection
+	clientLock  sync.RWMutex
+}
+
+type sqsConnection struct {
+	client        *sqs.SQS
+	config        *sqsConfig
+	queueURL      string
+	ctx           context.Context
+	ctxCancelFunc func() // used when we are destroying the connection or when there is an error
+}
+
+type sqsConfig struct {
+	event.SubscribeConfigDefaults
+	CreateQueue     bool   `mapstructure:"create_queue"`
+	AccessKeyID     string `mapstructure:"access_key_id"`
+	SecretAccessKey string `mapstructure:"secret_access_key"`
+	Region          string `mapstructure:"region"`
+	QueueName       string `mapstructure:"queue_name"`
+	QueueURL        string `mapstructure:"queue_url"`
+}
+
+func newClient(sconfig *sqsConfig) (*sqs.SQS, error) {
+	var options []awsutil.Option
+	if sconfig.AccessKeyID != "" && sconfig.SecretAccessKey != "" {
+		options = append(options, awsutil.WithAccessKey(sconfig.AccessKeyID))
+		options = append(options, awsutil.WithSecretKey(sconfig.SecretAccessKey))
+	}
+	if sconfig.Region != "" {
+		options = append(options, awsutil.WithRegion(sconfig.Region))
+	}
+	options = append(options, awsutil.WithEnvironmentCredentials(true))
+	credConfig, err := awsutil.NewCredentialsConfig(options...)
```

Comment:
I could be reading it wrong but it doesn't look like this function supports `WithEnvironmentCredentials`, and there might be some other supported options we're not consuming. Maybe worth pinging @austingebauer or @maxcoulombe to get some input on consistency as I'm not an expert.

Reply:
That's right that `NewCredentialsConfig` doesn't support `WithEnvironmentCredentials` from what I'm seeing. I'm not exactly sure what the motivation for that was. The only place the option is applied is within `GenerateCredentialChain`. I think you can also pass it through via `GetSession`.

I wish `awsutil` was more opinionated to help with consistency. A good start that's consistent with most AWS integrations would be to make sure the following credentials providers are in the chain (in order):

1.`credentials.StaticProvider`
2. `credentials.EnvProvider`
3. `credentials.SharedCredentialsProvider`

I do think it'd be worth checking out [`awsutil/v2`](https://github.com/hashicorp/go-secure-stdlib/blob/main/awsutil/go.mod) to get ahead of a potential future migration.

Reply:
For some reason I thought we couldn't import `awsutil/v2` in Vault until we moved everything over? So I was mostly copying what other users did.

Reply:
Oh, now I remember the problem -- `awsutil/v2` only provides clients for IAM and STS, and almost everything I would need to create an SQS client is private to the package.

So, I'll first need to setup `awsutil/v2` to provide an SQS client, and maybe make it more useful for non-IAM/STS use cases.

Reply:
Hmm. I might be able to use v2 without modifying `awsutil`. Sorry for all of the back and forth. :)